File: src/main/java/net/Chidoziealways/everythingjapanese/structure/custom/hellTemple/HellTempleStructurePieces.java
package net.Chidoziealways.everythingjapanese.structure.custom.hellTemple;

import com.google.common.collect.ImmutableList;
import com.mojang.datafixers.util.Pair;
import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.Chidoziealways.everythingjapanese.structure.ModProcessorLists;
import net.Chidoziealways.everythingjapanese.structure.ModStructuresR;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Holder;
import net.minecraft.core.HolderGetter;
import net.minecraft.core.registries.Registries;
import net.minecraft.data.worldgen.AncientCityStructurePools;
import net.minecraft.data.worldgen.BootstrapContext;
import net.minecraft.data.worldgen.Pools;
import net.minecraft.data.worldgen.ProcessorLists;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.ChunkPos;
import net.minecraft.world.level.StructureManager;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.chunk.ChunkGenerator;
import net.minecraft.world.level.levelgen.structure.BoundingBox;
import net.minecraft.world.level.levelgen.structure.StructurePiece;
import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
import net.minecraft.world.level.levelgen.structure.pools.StructurePoolElement;
import net.minecraft.world.level.levelgen.structure.pools.StructureTemplatePool;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessorList;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;

public class HellTempleStructurePieces {
    public static final ResourceKey<StructureTemplatePool> START = Pools.createKey(ResourceLocation.fromNamespaceAndPath(EverythingJapanese.MOD_ID, "hell_temple"));

    public static void bootstrap(BootstrapContext<StructureTemplatePool> context) {
        HolderGetter<StructureProcessorList> holdergetter = context.lookup(Registries.PROCESSOR_LIST);
        Holder<StructureProcessorList> holder = holdergetter.getOrThrow(ModProcessorLists.HELL_TEMPLE_START_DEGRADATION);
        HolderGetter<StructureTemplatePool> holdergetter1 = context.lookup(Registries.TEMPLATE_POOL);
        Holder<StructureTemplatePool> holder1 = holdergetter1.getOrThrow(Pools.EMPTY);
        context.register(
                START,
                new StructureTemplatePool(
                        holder1,
                        ImmutableList.of(
                                Pair.of(StructurePoolElement.single("everythingjapanese:hell_temple/hell_temple", holder), 1)
                        ),
                        StructureTemplatePool.Projection.RIGID
                )
        );
        HellTempleStructurePools.bootstrap(context);
    }

}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/structure/custom/hellTemple/HellTempleStructurePieces.kt
package net.Chidoziealways.everythingjapanese.structure.custom.hellTemple

import com.google.common.collect.ImmutableList
import com.mojang.datafixers.util.Pair
import net.Chidoziealways.everythingjapanese.EverythingJapanese
import net.Chidoziealways.everythingjapanese.MOD_ID
import net.Chidoziealways.everythingjapanese.structure.ModProcessorLists
import net.minecraft.core.Holder
import net.minecraft.core.registries.Registries
import net.minecraft.data.worldgen.BootstrapContext
import net.minecraft.data.worldgen.Pools
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.levelgen.structure.pools.StructurePoolElement
import net.minecraft.world.level.levelgen.structure.pools.StructureTemplatePool
import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessorList
import java.util.function.Function

object HellTempleStructurePieces {
    val START: ResourceKey<StructureTemplatePool?> =
        Pools.createKey(ResourceLocation.fromNamespaceAndPath(MOD_ID, "hell_temple/stairs"))

    fun bootstrap(context: BootstrapContext<StructureTemplatePool?>) {
        val holdergetter = context.lookup<StructureProcessorList?>(Registries.PROCESSOR_LIST)
        val holder: Holder<StructureProcessorList?> =
            holdergetter.getOrThrow(ModProcessorLists.HELL_TEMPLE_START_DEGRADATION)
        val holdergetter1 = context.lookup<StructureTemplatePool?>(Registries.TEMPLATE_POOL)
        val holder1: Holder<StructureTemplatePool?> = holdergetter1.getOrThrow(Pools.EMPTY)
        context.register(
            START,
            StructureTemplatePool(
                holder1,
                ImmutableList.of<Pair<Function<StructureTemplatePool.Projection?, out StructurePoolElement?>?, Int?>?>(
                    Pair.of<Function<StructureTemplatePool.Projection?, out StructurePoolElement?>?, Int?>(
                        StructurePoolElement.single("everythingjapanese:hell_temple/hell_stairs", holder),
                        1
                    )
                ),
                StructureTemplatePool.Projection.RIGID
            )
        )
        HellTempleStructurePools.bootstrap(context)
    }
}
